pkg/ottl/parser.go

```diff
@@ -20,28 +20,63 @@ import (
 	"go.uber.org/multierr"
 )
 
+// Statement holds a top level statement for processing telemetry data. A Statement is a combination of a function
+// invocation and the expression to match telemetry for invoking the function.
+type Statement[K any] struct {
+	Function  ExprFunc[K]
+	Condition BoolExpressionEvaluator[K]
```

Comment:
I think we should make the `Condition` an interface, names can be changed:

```bool
type Condition[K] interface {
  Eval(ctx context.Context, tCtx K) (bool, error)
}
```

Reply:
If not an interface, make it a struct with private members that exposes a `Eval` func. This way we can extend the functionality in the future, as defined right now is harder.

Reply:
I think using an interface exposing an `Eval` function would also help with transformation statements, since it would allow encapsulating conditionally running the function based on the `where` clause, as well as handling any new components of a transformation statement in the future. If we do this though, I think it would make sense to have the `Eval` function for a transformation statement return information on whether the telemetry item was matched (i.e. did the function run) and the return value from the function evaluation. This would allow components using the OTTL to take actions based on the result of the statement evaluation.

Reply:
Agreed. I will implement this feedback in its own PR, as it feels independent of any "condition parsing".

Reply:
@bogdandrutu @evan-bradley https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14869